pandas/tests/reshape/merge/test_merge.py

```diff
@@ -1318,6 +1318,44 @@ def test_merge_right_index_right(self):
         result = left.merge(right, left_on="key", right_index=True, how="right")
         tm.assert_frame_equal(result, expected)
 
+    @pytest.mark.parametrize("how", ["left", "right"])
+    def test_merge_preserves_row_order(self, how):
+        # GH 27453
+        a = [2, 5, 3, 5]
+        df1 = pd.DataFrame({"A": a, "B": [8, 2, 4, 1]})
+        df2 = pd.DataFrame({"A": a, "B": [7, 1, 3, 0]})
+
+        result = df1.merge(df2[["A", "B"]], on=["A", "B"], how=how)
+        expected = pd.DataFrame({"A": a})
+        if how == "right":
+            expected["B"] = df2["B"]
+        else:
+            expected["B"] = df1["B"]
+        tm.assert_frame_equal(result, expected)
+
+        left_df = pd.DataFrame({"colors": ["blue", "red"]}, index=pd.Index([0, 1]))
```

Comment:
either reduce the examples here (they look like 3x of the same one), or parameterize over them.